src/Backend/JSON.idr

```diff
@@ -0,0 +1,124 @@
+module Backend.JSON
+
+import Types
+import Typedefs
+
+import Backend
+import Backend.Utils
+
+import Control.Monad.State
+import Language.JSON
+import Text.PrettyPrint.WL
+
+import Data.Vect
+
+%default total
+%access public export
+
+||| A definition of a JSON schema.
+record JSONDef where
+  constructor MkJSONDef
+  name   : Name
+  schema : JSON
+
+||| Given `name`, create `[name0, name1, ..., namek]`
+nary : Name -> Vect k Name
+nary name = map ((name ++) . show . finToNat) range
+
+||| Reference a definition.
+defRef : Name -> JSON
+defRef name = JObject [("$ref", JString $ "#/definitions/" ++ name)]
+
+mutual
+  ||| Generate a schema for a `TDef 0`. The schema will be unusable without the definitions for the types referenced in the `TDef 0`.
+  makeSubSchema : TDef 0 -> JSON
+  makeSubSchema T0             = defRef "emptyType"
+  makeSubSchema T1             = defRef "singletonType"
+  makeSubSchema (TSum ts)      = disjointSubSchema $ zip (nary "case") ts
+  makeSubSchema (TProd ts)     = productSubSchema (nary "proj") ts
+  makeSubSchema (TName name _) = defRef name
+  makeSubSchema (TMu name _)   = defRef name
+  
+  ||| Generate a schema that matches exactly one of the supplied schemas, which must be wrapped in its corresponding name.
+  disjointSubSchema : Vect k (Name, TDef 0) -> JSON
+  disjointSubSchema cases = JObject [("oneOf", JArray . toList $ map makeCase cases)]
+    where makeCase : (Name, TDef 0) -> JSON
+          makeCase (name, td) = JObject
+                                  [ ("type", JString "object")
+                                  , ("required", JArray [JString name])
+                                  , ("additionalProperties", JBoolean False)
+                                  , ("properties", JObject [(name, assert_total $ makeSubSchema td)])
+                                  ]
+
+  ||| Generate a schema that requires all of the supplied schemas to match, with each being wrapped in its corresponding name.
+  productSubSchema : Vect k Name -> Vect k (TDef 0) -> JSON
+  productSubSchema names tds = JObject
+                                 [ ("type", JString "object")
+                                 , ("required", JArray . toList $ map JString names)
+                                 , ("additionalProperties", JBoolean False)
+                                 , ("properties", JObject . toList $ zip names (map (assert_total makeSubSchema) tds))
+                                 ]
+
+||| Only perform an action if a name is not already present in the state. If the action is performed, the name will be added.
+ifNotPresent : Name -> State (List Name) (List (String, JSON)) -> State (List Name) (List (String, JSON))
+ifNotPresent name gen = do
+    st <- get
+    if name `Prelude.List.elem` st
```

Comment:
I think it's usually enough to disambiguate with the last part of the package, i.e. `List.elem`